File: Ahorn/entities/SpiralStaircase.jl
module SpiralStaircase

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/SpiralStaircase" SpiralStaircase(x::Integer, y::Integer, clockwise::Bool=true, maxSpeed::Number=2.0, platformCount::Integer=8, radius::Integer=100, rotationSpeed::Number=0.5)

const placements = Ahorn.PlacementDict(
    "clockwise" => Ahorn.EntityPlacement(SpiralStaircase),
    "counterclockwise" => Ahorn.EntityPlacement(SpiralStaircase)
)

function Ahorn.selection(entity::SpiralStaircase)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SpiralStaircase, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/spiralStaircase.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local spiralStaircase = {}

spiralStaircase.name = "MaggyHelper/SpiralStaircase"
spiralStaircase.depth = -100
spiralStaircase.fieldInformation = {
    rotationSpeed = {
        minimumValue = 0
    },
    maxSpeed = {
        minimumValue = 0
    },
    platformCount = {
        minimumValue = 1
    },
    radius = {
        minimumValue = 0
    }
}
spiralStaircase.placements = {
    {
        name = "clockwise",
        data = {
            rotationSpeed = 0.5,
            maxSpeed = 2.0,
            platformCount = 8,
            radius = 100,
            clockwise = true
        }
    },
    {
        name = "counterclockwise",
        data = {
            rotationSpeed = 0.5,
            maxSpeed = 2.0,
            platformCount = 8,
            radius = 100,
            clockwise = false
        }
    }
}

function spiralStaircase.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/spiral_staircase/center", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 0.5)
    return sprite
end

function spiralStaircase.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 12, 24, 24)
end

return spiralStaircase
